src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeFactory.cs

```diff
@@ -0,0 +1,421 @@
+
+            static Evaluation? VisitNode(IOperation operation, Evaluation? input)
+            {
+                return operation switch
+                {
+                    IMemberReferenceOperation op when IsImplicitThisOrBaseMemberAccess(op) => null,
+                    ILocalReferenceOperation op => new Variable(input: null, op.Local, op.Syntax),
+                    IFieldReferenceOperation op => new MemberEvaluation(VisitInput(op.Instance, input), op),
+                    IPropertyReferenceOperation op when op.Property.IsIndexer => VisitIndexerReference(op),
+                    IPropertyReferenceOperation op => VisitNullableTypeMemberAccess(op) ??
+                                                      new MemberEvaluation(VisitInput(op.Instance, input), op),
+                    IConditionalAccessOperation op => VisitNode(op.WhenNotNull, VisitInput(op.Operation, input)),
+                    IConversionOperation op when op.Conversion.IsUserDefined => null,
+                    IConversionOperation op when op.Conversion.IsImplicit => VisitInput(op.Operand, input),
+                    IConversionOperation op => new Type(VisitInput(op.Operand, input), op.Type, op.Syntax),
+                    IConditionalAccessInstanceOperation => input,
+                    _ => null,
+                };
+            }
+
+            static bool IsImplicitThisOrBaseMemberAccess(IMemberReferenceOperation operation)
+            {
+                // We cannot use `base` as an standalone expression. So if we encounter `base.x.y.z` we
+                // will walk down until the very last member reference, but not the `base` node itself,
+                // resulting to the following graph:
+                //
+                //  base.x <- y <- z
+                //
+                return operation.Instance is IInstanceReferenceOperation instance &&
+                       instance.ReferenceKind == InstanceReferenceKind.ContainingTypeInstance &&
+                       (instance.IsImplicit || instance.Syntax.IsKind(SyntaxKind.BaseExpression));
+            }
+
+            static Evaluation? VisitIndexerReference(IPropertyReferenceOperation op)
+            {
+                // If this is the `ITuple.this[int]` indexer with a constant argument,
+                // record it as a separate node. We'll try to rewrite it as a positional pattern.
+
+                var indexer = op.Property;
+                Debug.Assert(indexer.IsIndexer);
+                var containingType = indexer.ContainingType;
+                if (containingType != null &&
+                    containingType.Equals(GetITupleType(op)) &&
+                    op.Arguments.Length == 1 &&
+                    IsConstant(op.Arguments[0].Value, out var constant) &&
+                    constant.Type.SpecialType == SpecialType.System_Int32)
+                {
+                    return new IndexEvaluation(VisitInput(op.Instance), indexer, (int)constant.ConstantValue.Value, op.Syntax);
```

Comment:
you have a lot of transformations that match a shape and return a new shape.  i would personally like comments showing what the transformation is to help doc things.